src/checkers/inference/solver/GeneralSolver.java

```diff
@@ -0,0 +1,407 @@
+            this.useGraph = true;
+        } else {
+            this.useGraph = false;
+        }
+
+        if (backEndType.equals(BackEndType.LOGIQL)) {
+            // Configure solving strategy.
+            this.solveInParallel = false;
+        } else if (solveInParallel == null || solveInParallel.equals(Constants.TRUE)) {
+            this.solveInParallel = true;
+        } else {
+            this.solveInParallel = false;
+        }
+
+        if (collectStatistic == null || collectStatistic.equals(Constants.FALSE)) {
+            // Configure statistic collection.
+            this.collectStatistic = false;
+        } else if (collectStatistic.equals(Constants.TRUE)) {
+            this.collectStatistic = true;
+        }
+
+        // Sanitize the configuration if it needs.
+        sanitizeConfiguration();
+        System.out.println("Configuration: \nback end type: " + this.backEndType.simpleName + "; \nuseGraph: "
+                + this.useGraph + "; \nsolveInParallel: " + this.solveInParallel + ".");
+    }
+
+    protected void configureLattice(QualifierHierarchy qualHierarchy) {
+        LatticeBuilder latticeBuilder = new LatticeBuilder();
+        lattice = latticeBuilder.buildLattice(qualHierarchy);
+    }
+
+    protected TwoQualifiersLattice createTwoQualifierLattice(AnnotationMirror top, AnnotationMirror bottom) {
+        LatticeBuilder latticeBuilder = new LatticeBuilder();
+        return latticeBuilder.buildTwoTypeLattice(top, bottom);
+    }
+
+    /**
+     * This method creates a ConstraintSerializer, which can deliver the
+     * constraint/slot to the real serializer. If a costume serialization logic
+     * is needed, user can have a subclass of ConstraintSerializer, override the
+     * serialize method with the costume logic, and then override this method
+     * with returning the instance of new subclass.
+     * 
+     * @param backEndType
+     * @param lattice
+     * @return A deliverer Serializer.
+     */
+    protected Serializer<?, ?> createSerializer(BackEndType backEndType, Lattice lattice) {
+        return new ConstraintSerializer<>(backEndType, lattice);
+    }
+
+    protected ConstraintGraph generateGraph(Collection<Slot> slots, Collection<Constraint> constraints,
+            ProcessingEnvironment processingEnvironment) {
+        GraphBuilder graphBuilder = new GraphBuilder(slots, constraints);
+        ConstraintGraph constraintGraph = graphBuilder.buildGraph();
+        return constraintGraph;
+    }
+
+    protected BackEnd<?, ?> createBackEnd(BackEndType backEndType, Map<String, String> configuration,
+            Collection<Slot> slots, Collection<Constraint> constraints,
+            QualifierHierarchy qualHierarchy, ProcessingEnvironment processingEnvironment,
+            Lattice lattice, Serializer<?, ?> defaultSerializer) {
+
+        BackEnd<?, ?> backEnd = null;
+
+        try {
+            Class<?> backEndClass = Class.forName(backEndType.fullyQualifiedName + "BackEnd");
```

Comment:
Just have a question: what is the difference between directly make a sub-class of `MaxSatSerializer` and a sub-class of `ConstraintSerializer`?

 The only difference I can see is a sub-class of `ConstraintSerializer` can achieve multi-backends serialization (e.g. having MaxSat and LogiQL serialization implemented in the same overriden method), but this is still makes code complicate (i.e. we make different backend serialization into one method).

 However, if we want to separate serializer implementation for different backends, then there is no difference to just make sub-classes to concrete backends, and then in the `createSerializer()` method, we can have:

```java
protected Serializer<?, ?> createSerializer(BackEndType backEndType, Lattice lattice) {
        switch(backEndType) {
        case MAX_SAT: return new customMAXSATSerialzer(); break;
        ....
        default: throws new Exception("this backend type is neither implemented nor supported!");
        }
    }
```

In the `GeneralSolver`, we still initialize serializer by just one line code `BackEndType.getNewDefaultSerializer()`.